# Assignment 3

*Part of the course:
Machine Learning (code: INFOB3ML), fall 2023, Utrecht University*

Total points: 10 (100%)

Deadline: Wednesday 18 October, 23:59

**Write your names and student numbers here: Joel van de Hoeve 5280273 & Thijmen van der Meijden 1670786**

Submit one ipynb file per pair.

**Before you submit, click Kernel > Restart & Run All to make sure you submit a working version of your code!**



## The BoAW Approach for emotion classification of utterances
In this programming assignment, you will be implementing Bag-of-Audio-Words (BoAW) method for acoustic low level descriptor (LLD) representation over utterances.
The acoustic LLDs per utterance are extracted and provided to you in separate files.
Your task is to implement the BoAW approach and the component methods, followed by classification using existing python libraries.

In a nutshell, you will be implementing the following functions (use of an existing library/package like `sklearn` is not allowed for this part):
1. Principal Component Analysis (PCA)
2. K-Means algorithm
3. BoAW representation function

You will then experiment with different combinations of PCA dimensions (`p_pca`) and K-Means components (`K`), and optimize said combinations using a Support Vector Machine (SVM) classifier.
You are given data of 16 speakers, where:
- The data of the first 8 speakers will be used as training set
- The data of speakers 9-12 will be used as validation set and
- The data of speakers 13-16 will be used as the test set.

The training set will be used to learn PCA and K-Means model, as well as to train the SVM classifiers using the BoAW representations as features and the emotions of the speakers as responses.
The hyper-parameter `C` of the classifiers as well as the number of principal components (`p_pca`) and cluster components (`K`) will be optimized on the validation set.
The top model (resulting from the best combination of `p_pca`, `K` and `C` hyper-parameters) will be applied to the test set.

### The BoAW approach
In the BoAW approach (can also be applied to visual descriptors) we train a K-Means model using the combined LLDs from all trained set utterances.
Since K-Means assumes no correlations in the data, and partly for reducing dimensionality, we first apply PCA to LLDs.
The BoAW representation amounts to assigning each LLD to the nearest K-Means cluster and computing a K-dimensional histogram of the accumulated instances around each mean.
This means that each utterance will be "represented" by a histogram, where each entry is the number of LLD sequences in the utterance that are assigned to a certain cluster.
Normalization of the histogram (so that each entry is a frequency as opposed to an absolute number) is possible.

### Note
The last part of this assignment contains a lot of code that you do not need to (and should not) modify. However, it is important that you understand what is going on there, otherwise it will be difficult to fill-in the blanks when we ask you to.

The only places where you need to write code say "YOUR CODE HERE" explicitly.

### Required packages:
- numpy
- pandas
- scipy
- sklearn


### Required Dataset
You can download the zipped folder containing the extracted acoustic LLDs from subset of RAVDESS dataset from:

https://surfdrive.surf.nl/files/index.php/s/EeJxsea0CfCn17S

Unzip it in the same folder as this python notebook to ensure smooth working of the scripts we have prepared.

In [80]:
# DO NOT MODIFY
import numpy as np

### Step 1: Implementing PCA (Graded 20%)
Define a function `mypca(X)` with the following signature:
 - Input  `X`: N x D dimensional dataset
 - Output: `(W, lambda, mu)`
     + `W`: D x D eigenvalue sorted PCA projection matrix (i.e. first column is the first principal component)
     + `lambdas`: Dx1 vector of sorted eigenvalues (sorted in _descending_ order)
     + `mu`: Dx1 the mean of X (should be calculated and removed from the data before subsequent projection using W)
     
Hints:
 1. Recall that you should start by ensuring that the data has zero mean.
 2. You may want to make use of the numpy function `np.cov`. Be careful with what matrix you feed into this function. (Extra hint: look at the description `rowvar` argument of this function in the documentation).
 3. You will want to sort the eigenvalues in descending order, but numpy's sort method can only sort in ascending order. You may want to consider using `np.argsort` and using the fact that the ascending sorting indices for `-v` and the descending sorting indices for `v` are the same.
 4. Other numpy functions you may want to use: `np.mean`, `np.linalg.eig`.

In [81]:
def mypca(X):
    mu = np.mean(X, axis=0)
    X2 = X - mu

    cov_matrix = np.cov(X2, rowvar=False)

    eigenvalues, eigenvectors = np.linalg.eig(cov_matrix)
    sorted_index = np.argsort(eigenvalues)[::-1]

    lambdas = eigenvalues[sorted_index]
    W = eigenvectors[:, sorted_index]
    return W, lambdas, mu

In [82]:
# TEST
np.random.seed(0)
X = np.random.rand(4, 3) # Three features; four samples.
W, lambdas, mu = mypca(X)
print(f"W = {W}")
print(f"\nlambdas = {lambdas}")
print(f"\nmu = {mu}")

assert W.shape == (3, 3), "Wrong shape of array W!"
assert len(lambdas) == 3, "Wrong length of vector lambdas!"
assert len(mu) == 3, "Wrong length of vector mu!"

W = [[-0.19056484  0.29124074  0.9374774 ]
 [ 0.75580654 -0.56588686  0.32943669]
 [ 0.62645153  0.7713306  -0.11228352]]

lambdas = [0.0568462  0.02424662 0.00291952]

mu = [0.47868135 0.70558555 0.68530379]


### Step 2: Implementing K-Means (Graded 40%)
Define a function `mykmeans(X, K, maxiter=200, epsilon=0.0001)` with the following signature:
 - Input
     + `X`: N x D dimensional dataset
     + `K`: Hyper-parameter (number of clusters) of K-Means
     + `maxiter`: maximum number of iterations (default 200)
     + `epsilon`: algorithm will stop when error improvement < epsilon (or number of iterations surpasses `maxiter`) (default 200)
 - Output: `(M, A)`
     + `M`: K x D dimensional matrix of means
     + `A`: N x 1 dimensional cluster assignments
     
Notes:
 1. It is not necessary but you may want to test your function in a simple example that is easy to visualize, like the ones in the book. In that case, you'll need to come up with it yourself. If you do so, I would advise doing so in a separate Python file and if you do it in this notebook _do not leave that extra code in, or you'll be penalized_.
 2. You will need to compute distances between samples `X[n]` and mean vectors `M[k]`. You can use the function `dist` given below for that. Do not modify it.

In [83]:
# DO NOT MODIFY
from scipy.spatial.distance import cdist
# Hint: you will need this function in your K-Means (mykmeans) and BoAW (myboaw) function implementations
def dist(X, M):
    """Computes the matrix with entries dist_matrix_nk = euclidean distance between X[n] and M[k] efficiently."""
    return cdist(X,M,metric='euclidean')

In [201]:
def mykmeans(X, K, maxiter=200, epsilon=0.0001):
    np.random.seed(1)
    N, D = X.shape
    # initialize model/output variables
    initial_indices = np.random.choice(N, K) # sample random K out of N instances to use as initial means
    M = X[initial_indices, :] # (K, D) matrix
    min_indices = np.zeros((N, 1))
    error_log = np.zeros((maxiter,1)) # You can save the errors here, to then use for computing delta
    iteration = 0
    cont = 1 # set to 0 when convergence criteria is met

    print("Running k-means...")
    while cont:
        dist_matrix = dist(X,M)
        min_indices = np.argmin(dist_matrix, axis=1)

        M2 = np.array([X[min_indices == k].mean(axis=0) if X[min_indices == k].shape[0] > 0 else M[k] for k in range(K)])

        if iteration == 0:
            delta = 100000 # Something very high
        else:
            delta = np.mean(np.linalg.norm(M2 - M, axis=1))
            
        M = M2
        error_log[iteration] = delta

        if (delta <= epsilon or iteration == maxiter):
            cont=0

        assert not np.isnan(delta), "delta is nan!"

        print(f"iteration: {iteration}, delta: {delta}")
        iteration = iteration + 1

    # Final update of the indices specifying the clusters for each sample
    dist_matrix = dist(X,M)
    min_indices = np.argmin(dist_matrix, axis=1)
    
    A = min_indices # Just renaming

    return M, A

In [202]:
# TEST
np.random.seed(0)
X = np.random.randint(1, 10, (12, 2)) #  two features; 12 samples.
K = 3
mykmeans(X, K)

Running k-means...
iteration: 0, delta: 100000
iteration: 1, delta: 0.8367721822455686
iteration: 2, delta: 0.8485990377670415
iteration: 3, delta: 0.0


(array([[6.75      , 8.5       ],
        [4.4       , 3.        ],
        [8.66666667, 3.66666667]]),
 array([1, 1, 2, 1, 0, 0, 2, 0, 0, 1, 2, 1], dtype=int64))

### Step 3: Implementing Bag of Audio Words (BoAW) representation (Graded 20%)
Define a function `myboaw(X_LLD, M)` which assigns the set of LLDs in the first argument to the nearest mean using the matrix of means given in the second argument.
The function should have the following signature:
 - Input:
     + `X_LLD`: N x D matrix of LLDs from each utterance, where N is the number of frames / instances and D is the (reduced) LLD dimensionality
     + `M`: K x D matrix of means (code vectors) outputed from K-Means
 - Output:
     + `boaw_rep`: K x 1 matrix with the _normalized_ boaw representation for each cluster k in each entry.
     
Hint:
 - After obtaining the K x 1 vector of unnormalized boaw representations (whose entries are integers), you can get the normalized version by simple dividing the unnormalized vector by the sum over all the entries of the unnormalized boaw vector.

In [86]:
def myboaw(X_LLD, M):
    K, _ = M.shape

    dist_matrix = dist(X_LLD, M)         
    clusterings = np.argmin(dist_matrix, axis=1)

    boaw_unnormalised = np.bincount(clusterings, minlength=K)
    boaw_rep = boaw_unnormalised / np.sum(boaw_unnormalised)

    return boaw_rep

In [166]:
# TEST
np.random.seed(0)
X = np.random.randint(1, 10, (12, 2)) #  two features; 12 samples.
K = 3
M, _ = mykmeans(X, K)
print(myboaw(X, M))

Running k-means...
iteration: 0, delta: 100000
iteration: 1, delta: 1.4837962962962965
iteration: 2, delta: 1.087044595616024
iteration: 3, delta: 0.0
[0.33333333 0.41666667 0.25      ]


## Step 4: BoAW representation in action (graded 20%)

### Step 4.1: Data ingestion (0% - we do it for you!)
Read the data and place them in an appropriate data structure such that LLD sequence coming from each file (=utterance) is contained separately.

NOTES:
 1. Even though you do not need to write code for this part, make sure you understand what is going on, otherwise you will not comprehend the next steps.
 2. I would advise you to put the RAVDESS directory in the directory where this notebook is being run from - otherwise you will need to modify the variable `base_folder` below, as is written there.

In [88]:
# DO NOT MODIFY

import glob
import os
import pandas as pd


# Each file will have some data and metadata we will want to record. We do so in using a class:
class FileData():
    def __init__(self, speakerid, emoid, LLDseq, LLDcount):
        self.speakerid = speakerid
        self.emoid = emoid
        self.LLDseq = LLDseq
        self.LLDcount = LLDcount


def ingest_ravdess_data(base_folder):
    """Build list of FileData instances, one per file/utterance."""
    LLD_folder = 'LLD_CSV_IS13/'
    list_actors_dirs = sorted(glob.glob(os.path.join(base_folder, LLD_folder, 'Actor*')))
    #print(list_actors_dirs)

    cnt = 0
    LLD_data = [] # To be populated by instances of FileData

    # The metadata can be obtained from the provided metadata.csv or from the
    # filename directly. We'll do the latter.

    for actor_folder in list_actors_dirs:
        list_actorLLDfiles = sorted(glob.glob(os.path.join(actor_folder, '*.csv')))
        #print(list_actorLLDfiles)
        for actor_file in list_actorLLDfiles:
            #print(f"file: {actor_file}")
            cnt = cnt + 1
            tmp_data_og = pd.read_csv(actor_file, sep=';')
            tmp_data = tmp_data_og.iloc[:, 2:]
            # if cnt == 1: # Visualise data on first iteration
            #     print(tmp_data.shape)
            #     print(tmp_data.iloc[:5, :4])
            #     cols = tmp_data.columns
            #     #print(f"Column names: {["\n\t" + f"{col}" for col in cols]}")
            #     print("\nColumns:")
            #     for col in cols:
            #         print("\t" + f"{col}")
            actor_file_wo_ext = os.path.splitext(actor_file)[0] # Remove extension from file name
            filecodes = str.split(actor_file_wo_ext,'-') # list with the filecodes, for example ['03', '02', ...]
            emoid = int(filecodes[2]) # the third filecode corresponds to the emotion id
            #print(f"emoid: {emoid}")
            speakerid = int(filecodes[-1]) # the last filecode corresponds to the actor/speaker id
            LLD_data.append(FileData(speakerid, emoid, tmp_data, tmp_data.shape[0]))
    assert len(LLD_data) == 704, "Something went wrong! There are 704 csv files in the RAVDESS dataset, but len(LLD_data) != 704."

    return LLD_data

In [89]:
# ONLY MODIFY IF BASE-FOLDER IS DIFFERENT!
base_folder = './RAVDESS/' # Assuming that the RAVDESS directory is in the directory where this notebook is being run from.
LLD_data = ingest_ravdess_data(base_folder) # LLD_data is a data structure that keeps the acoustic LLD sequences for each audio file
# The goal of this assignment is to convert the (arbitrary length) LLD sequences into a fixed length utterance representation using Bag-of-Audio-Words approach

In [90]:
# DO NOT MODIFY

# In this part, each function does one of these things:
#  1. Generate train data table with one LLDseq (corresponding to one audio frame) per row, using the first few speakers only.
#  2. Generate the mask arrays (using the speakerids) that will be used in the rest of this notebook
#        for splitting the datasets in train, val and test sets.
#  3. Generate the response variable vectors for the train, val and test sets.

def build_ravdess_train_dataset(LLD_data, n_speakers_train=8):
    """Builds RAVDESS dataset from list of FileData corresponding to the LLD of the utterances.

    Input:
        LLD_data -- list of FileData instances, one for each utterance file.

    Output:
        all_train_LLDsequences -- dataframe of shape (N_seqs=total number of LLD seqs, number of LLD features),
                                    with one LLD sequence per row.

    """
    all_train_LLDsequences = pd.concat([file_data.LLDseq for file_data in LLD_data if file_data.speakerid <= n_speakers_train]) # Each LLDseq is one feature vector

    return all_train_LLDsequences


def generate_splitting_masks(LLD_data, n_speakers_train=8, n_speakers_val=4):
    all_speakerids = np.array([file_data.speakerid for file_data in LLD_data])

    # Masks to build dataset using speaker ids:
    mask_train = all_speakerids <= n_speakers_train
    mask_val = np.logical_and(n_speakers_train < all_speakerids, all_speakerids <= n_speakers_train + n_speakers_val)
    mask_test = n_speakers_train + n_speakers_val < all_speakerids

    return mask_train, mask_val, mask_test, all_speakerids

def prepare_ravdess_response_data(LLD_data, mask_train, mask_val, mask_test):
    all_emoids = np.array([file_data.emoid for file_data in LLD_data])

    # Response's part of the dataset
    emoids_train = all_emoids[mask_train]
    emoids_val = all_emoids[mask_val]
    emoids_test = all_emoids[mask_test]

    return emoids_train, emoids_val, emoids_test

In [91]:
# DO NOT MODIFY
# Make 0.5-0.25-0.25 train-val-test split using speakerid
# (i.e. 8 speakers for the train, 4 speakers for the val and 4 speakers for the test sets).
mask_train, mask_val, mask_test, _ = generate_splitting_masks(LLD_data)
emoids_train, emoids_val, emoids_test = prepare_ravdess_response_data(LLD_data, mask_train, mask_val, mask_test)
LLDs_train = build_ravdess_train_dataset(LLD_data)

In [92]:
# TEST:
print(LLDs_train.shape)
print(emoids_train[:44]) # Look at some responses

assert LLDs_train.shape[1] == 130, "There should be 130 features."

(162621, 130)
[1 1 1 1 2 2 2 2 2 2 2 2 3 3 3 3 3 3 3 3 4 4 4 4 4 4 4 4 5 5 5 5 5 5 5 5 6
 6 6 6 6 6 6 6]


### Step 4.2: Learn PCA projection (5%)
Learn a PCA projection (i.e. the eigenvectors and eigenvalues of the covariance matrix) using your own PCA implementation, namely `mypca`.

NOTE:
If there is a large variance across variances of the features then it is wise to z-normalize the data before applying PCA (using for example an instance of the `sklearn.preprocessing.StandardScaler` class).

Running the code below shows that indeed there is a huge difference in variances

In [93]:
# DO NOT MODIFY
data_var = LLDs_train.var()
print(f"Variances:\n{'-'*5}\n{data_var}\n{'-'*5}")
#ratio_extremes = np.max(data_var)/np.min(data_var)
var_of_variances = np.var(data_var)
#print(ratio_extremes)
print(f"Variance of the variances: {var_of_variances}")

Variances:
-----
F0final_sma                  23295.126736
voicingFinalUnclipped_sma        0.184585
jitterLocal_sma                  0.002086
jitterDDP_sma                    0.001889
shimmerLocal_sma                 0.010514
                                 ...     
mfcc_sma_de[10]                  6.408354
mfcc_sma_de[11]                  6.417294
mfcc_sma_de[12]                  5.963105
mfcc_sma_de[13]                  4.900797
mfcc_sma_de[14]                  4.299584
Length: 130, dtype: float64
-----
Variance of the variances: 5.4100025350023796e+26


In [94]:
# You now need to Z-normalize LLDs_train. Call the resulting variable X_zn. You can use sklearn.preprocessing.StandardScaler.
from sklearn.preprocessing import StandardScaler
# Instantiate scaler

scaler = StandardScaler()
# Fit scaler
X_zn = scaler.fit_transform(LLDs_train) 
# Finally, apply mypca to extract pca_W and pca_lambdas
pca_W, pca_lambdas, _ = mypca(X_zn)

In [95]:
# TEST:
print(X_zn.shape)
print(pca_W.shape)
print(pca_lambdas.shape)

assert X_zn.shape == (162621, 130), "Something went wrong!"
assert pca_W.shape == (130, 130), "Something went wrong!"
assert pca_lambdas.shape == (130,), "Something went wrong!"

(162621, 130)
(130, 130)
(130,)


### Step 4.3: Applying PCA and K-means (5%)

Project the training set using the first `p_pca` PCA components and train a K-Means model with K components.
Using a number `p_pca` of principal components equal to 120 and a number of clusters `K` of 256.

In [203]:
# DO NOT MODIFY
p_pca = 120
K = 256

We already have the pre-processed (z-normalized) training LLD data.

First, get PCA projection onto the (basis spanned by) the first p_pca eigenvectors.
Call the resulting matrix `Z_LLD_train`.

Hint: This function should not have more than a couple lines.

In [191]:
def project_onto_pcs(X, pca_W, p_pca):
    # Projecting coincides with matrix multiplication (if pca_W column vectors are normalized, which they are):
    X_projected = X.dot(pca_W[:, :p_pca])
    return X_projected

In [192]:
# DO NOT MODIFY
Z_LLD_train = project_onto_pcs(X_zn, pca_W, p_pca)

In [193]:
# TEST
print(Z_LLD_train.shape)
assert Z_LLD_train.shape == (162621, 120), "Something went wrong! Shape of Z_LLD_train is incorrect!"

(162621, 120)


In [194]:
# DO NOT MODIFY
# We will use only every fourth frame/LLD sequence, to reduce the computation time.

def subsample(X, n_subsample):
    X_subsample = X[np.arange(X.shape[0], step=n_subsample), :] # Use every n_sample'th point
    return X_subsample

n_subsample = 4 # Use every fourth frame
Z_LLD_train_subsample = subsample(Z_LLD_train, n_subsample)
print(Z_LLD_train_subsample.shape)

(40656, 120)


In [204]:
# Apply your kmeans implementation to the projected, subsampled, train data:
np.random.seed(0)
M, _ = mykmeans(Z_LLD_train_subsample, K)

Running k-means...
iteration: 0, delta: 100000
iteration: 1, delta: 1.4853194536069052
iteration: 2, delta: 1.0522894840014116
iteration: 3, delta: 0.8545235815997467
iteration: 4, delta: 0.7292941487224829
iteration: 5, delta: 0.6398216055408585
iteration: 6, delta: 0.6039399874692676
iteration: 7, delta: 0.4780239513337041
iteration: 8, delta: 0.3946065781465394
iteration: 9, delta: 0.3847282034798641
iteration: 10, delta: 0.3767501446923005
iteration: 11, delta: 0.3824625031632304
iteration: 12, delta: 0.39808426370921757
iteration: 13, delta: 0.34272415064020684
iteration: 14, delta: 0.3025265189681985
iteration: 15, delta: 0.26986832604909433
iteration: 16, delta: 0.2339084815741848
iteration: 17, delta: 0.2311520818546386
iteration: 18, delta: 0.24330338217228434
iteration: 19, delta: 0.18462894840477537
iteration: 20, delta: 0.1570275245144761
iteration: 21, delta: 0.14894002243841312
iteration: 22, delta: 0.13270682232083902
iteration: 23, delta: 0.1509799135015853
iteration: 2

### Step 4.4: BoAW encode each file using the loaded data (5%)
Using the result of the K-means method and LLD_data, use your implementation of BoAW to build one histogram per utterance/file.

In [196]:
# We can collect all boaw representations in one matrix. We will later split it!

def boaw_representations(LLD_data, scaler, pca_W, p_pca, K, M):
    "Generate matrix with one boaw representation (for each utterance) per row."
    N_files = len(LLD_data) # One sample per file/utterance, each sample being a histogram.
    #print(f"Number of utterances: {N_files}")
    X_boaw_all = np.zeros((N_files, K)) # Each row will have K entries, each with the frequency of one of the clusters in that utterance/audio file.

    for i in range(N_files):
        LLD_normalized = scaler.transform(LLD_data[i].LLDseq)
        LLD_projected = project_onto_pcs(LLD_normalized, pca_W, p_pca)
        n_boaw_rep = myboaw(LLD_projected, M)
        X_boaw_all[i] = n_boaw_rep

    return X_boaw_all

In [197]:
# DO NOT MODIFY
X_boaw_all = boaw_representations(LLD_data, scaler, pca_W, p_pca, K, M)
print(f"X_boaw_all shape: {X_boaw_all.shape}")
assert X_boaw_all.shape == (704, 256), "Something went wrong!"

X_boaw_all shape: (704, 256)


In [198]:
# DO NOT MODIFY
# And now we split the representations in train and val sets

def split_ravdess_filewise_boaw(X_boaw_all, mask_train, mask_val, mask_test):
    """Use the masks computed before to split X_boaw_all in train, val and test parts."""
    X_boaw_train = X_boaw_all[mask_train]
    X_boaw_val = X_boaw_all[mask_val]
    X_boaw_test = X_boaw_all[mask_test]

    return X_boaw_train, X_boaw_val, X_boaw_test

In [199]:
# DO NOT MODIFY
X_boaw_train, X_boaw_val, X_boaw_test = split_ravdess_filewise_boaw(
    X_boaw_all, mask_train, mask_val, mask_test
)
Y_train = emoids_train
Y_val = emoids_val
Y_test = emoids_test

In [200]:
# TEST
print(X_boaw_train.shape)
print(X_boaw_val.shape)
print(X_boaw_test.shape)
print(Y_train.shape)
print(Y_val.shape)
print(Y_test.shape)

(352, 256)
(176, 256)
(176, 256)
(352,)
(176,)
(176,)


### Step 4.5: (5%)
Use the sklearn implementation of LinearSVC (linear support vector machine) to optimize C using the validation set.

To optimize C, simply try C = 10^-6, 10^-5, 10^-4, 10^-3, 10^-2, 10^-1, 10^-0, and select the one giving the best accuracy.
Make sure to output best performance (i.e. accuracy) and corresponding C per BoAW representation.

In [219]:
from sklearn.svm import LinearSVC

C_set = 10.**np.arange(-6, 1) # the experimental set of SVM complexity values

def generate_performances(X_train, X_val, Y_train, Y_val, C_set):
    """Compute the performances of linear svc (with random state = 1) for each C in C_set.

    Output:
        performances -- (len(C_set), 1) array witha accuracy corresponding to C_set[i] on entry i.

    """
    performances = np.zeros((len(C_set), 1)) # Performace for a certain C will simply be average correct preds = accuracy
    for i in range(len(C_set)):
        C = C_set[i]
        ding = LinearSVC(C=C, random_state=1, dual='auto')

        ding.fit(X_train, Y_train)

        acc = ding.score(X_val, Y_val)
        performances[i] = acc
    return performances

def find_best_C(C_set, performances):
    """Determines the best C in C_set using the list performances of accuracies.

    Output:
        best_C_boaw -- float
        best_accuracy_boaw -- float
    """
    index = np.argmax(performances)
    best_C_boaw = C_set[index]
    best_accuracy_boaw = performances[index]

    return best_C_boaw, best_accuracy_boaw

In [220]:
# DO NOT MODIFY
performances = generate_performances(X_boaw_train, X_boaw_val, Y_train, Y_val, C_set)
best_C_boaw, best_accuracy_boaw = find_best_C(C_set, performances)

print(f"For K={K} and p_pca={p_pca}:")
print(f"\tBest accuracy: {best_accuracy_boaw}")
print(f"\tBest C: {best_C_boaw}")

For K=512 and p_pca=130:
	Best accuracy: [0.43181818]
	Best C: 0.001


### Step 4.5: Repeat steps 4.3 and 4.4 with different values of `p_pca` and `K` (0% - we do it for you!)

Use `p_pca` in {120, 130}.
Use `K` in {256, 512}.
(Therefore you'll have 4 BoAW representations in total).

Important note: since there will be hundreds of thousands of LLD vectors
it will be wise to subsample them before K-Means taking one every `n_subsample=4` frames.

In [207]:
# DO NOT MODIFY
def generate_boaw_representation(X_zn, p_pca, pca_W, K, n_subsample, scaler):
    X_projected = project_onto_pcs(X_zn, pca_W, p_pca)
    X_projected_subsample = subsample(X_projected, n_subsample)
    M, _ = mykmeans(X_projected_subsample, K)
    X_boaw_all = boaw_representations(LLD_data, scaler, pca_W, p_pca, K, M)
    #print(X_boaw_all.shape)
    return X_boaw_all

def test_hyperparameters(
    X_zn, Y_train, Y_val, p_pcas, pca_W, Ks, n_subsample, scaler, C_set
):
    X_boaw_all_dict = {}
    for p_pca in p_pcas:
        for K in Ks:
            print("-"*5 + f"\np_pca: {p_pca}, K: {K}")

            # Generate the representation
            X_boaw_all = generate_boaw_representation(
                X_zn, p_pca, pca_W, K, n_subsample, scaler
            )

            # Split boaw data
            X_boaw_train, X_boaw_val, X_boaw_test = split_ravdess_filewise_boaw(
                X_boaw_all, mask_train, mask_val, mask_test
            )

            # One list of performances per representation
            performances = generate_performances(
                X_boaw_train, X_boaw_val, Y_train, Y_val, C_set
            )

            # One best C and best accuracy per representation
            best_C_boaw, best_accuracy_boaw = find_best_C(C_set, performances)

            print(f"\nFor K={K} and p_pca={p_pca}:")
            print(f"\tBest accuracy: {float(best_accuracy_boaw)}")
            print(f"\tBest C: {best_C_boaw}")

            X_boaw_all_dict[(p_pca, K)] = X_boaw_all

    return X_boaw_all_dict

In [208]:
# DO NOT MODIFY
n_subsample = 4
p_pcas = [120, 130]
Ks = [256, 512]
X_boaw_all_dict = test_hyperparameters(
    X_zn, Y_train, Y_val, p_pcas, pca_W, Ks, n_subsample, scaler, C_set
)

-----
p_pca: 120, K: 256
Running k-means...
iteration: 0, delta: 100000
iteration: 1, delta: 1.4853194536069052
iteration: 2, delta: 1.0522894840014116
iteration: 3, delta: 0.8545235815997467
iteration: 4, delta: 0.7292941487224829
iteration: 5, delta: 0.6398216055408585
iteration: 6, delta: 0.6039399874692676
iteration: 7, delta: 0.4780239513337041
iteration: 8, delta: 0.3946065781465394
iteration: 9, delta: 0.3847282034798641
iteration: 10, delta: 0.3767501446923005
iteration: 11, delta: 0.3824625031632304
iteration: 12, delta: 0.39808426370921757
iteration: 13, delta: 0.34272415064020684
iteration: 14, delta: 0.3025265189681985
iteration: 15, delta: 0.26986832604909433
iteration: 16, delta: 0.2339084815741848
iteration: 17, delta: 0.2311520818546386
iteration: 18, delta: 0.24330338217228434
iteration: 19, delta: 0.18462894840477537
iteration: 20, delta: 0.1570275245144761
iteration: 21, delta: 0.14894002243841312
iteration: 22, delta: 0.13270682232083902
iteration: 23, delta: 0.1509

C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C


For K=256 and p_pca=120:
	Best accuracy: 0.4034090909090909
	Best C: 0.001
-----
p_pca: 120, K: 512
Running k-means...
iteration: 0, delta: 100000
iteration: 1, delta: 1.4010563096708328
iteration: 2, delta: 0.9156136097233867
iteration: 3, delta: 0.699442658893203
iteration: 4, delta: 0.5571447894405692
iteration: 5, delta: 0.5013048010636119
iteration: 6, delta: 0.49472777683571556
iteration: 7, delta: 0.4122150869470027
iteration: 8, delta: 0.36097553280800676
iteration: 9, delta: 0.3328393152144958
iteration: 10, delta: 0.31641451374282414
iteration: 11, delta: 0.2637463363517932
iteration: 12, delta: 0.24167901267740605
iteration: 13, delta: 0.1786453997787683
iteration: 14, delta: 0.15483809463144535
iteration: 15, delta: 0.15115234463037208
iteration: 16, delta: 0.13475548618461128
iteration: 17, delta: 0.12336023841703397
iteration: 18, delta: 0.1137456132800909
iteration: 19, delta: 0.08915290867004307
iteration: 20, delta: 0.09082373761046678
iteration: 21, delta: 0.08521570

C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C


For K=512 and p_pca=120:
	Best accuracy: 0.42045454545454547
	Best C: 0.0001
-----
p_pca: 130, K: 256
Running k-means...
iteration: 0, delta: 100000
iteration: 1, delta: 1.4852234952703915
iteration: 2, delta: 1.0504549636455045
iteration: 3, delta: 0.8544125890994145
iteration: 4, delta: 0.7211980518590393
iteration: 5, delta: 0.6459126330085672
iteration: 6, delta: 0.6060193995389382
iteration: 7, delta: 0.49643239613324086
iteration: 8, delta: 0.39158620995956855
iteration: 9, delta: 0.37920426627675813
iteration: 10, delta: 0.36864557858982466
iteration: 11, delta: 0.340144401059989
iteration: 12, delta: 0.347546930704361
iteration: 13, delta: 0.3233640069827672
iteration: 14, delta: 0.3159552806182771
iteration: 15, delta: 0.2704699975245567
iteration: 16, delta: 0.2113571993052169
iteration: 17, delta: 0.2182846857481633
iteration: 18, delta: 0.21143154728164382
iteration: 19, delta: 0.17377317982407153
iteration: 20, delta: 0.19964266720875384
iteration: 21, delta: 0.1837988287

C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C


For K=256 and p_pca=130:
	Best accuracy: 0.4034090909090909
	Best C: 0.01
-----
p_pca: 130, K: 512
Running k-means...
iteration: 0, delta: 100000
iteration: 1, delta: 1.4028019879023914
iteration: 2, delta: 0.9210870449368416
iteration: 3, delta: 0.7003748410375121
iteration: 4, delta: 0.5566905144016674
iteration: 5, delta: 0.4989120591085026
iteration: 6, delta: 0.492187617749635
iteration: 7, delta: 0.4158032188992583
iteration: 8, delta: 0.3713212660199481
iteration: 9, delta: 0.34603843634659354
iteration: 10, delta: 0.3134037799610784
iteration: 11, delta: 0.2743071095733838
iteration: 12, delta: 0.23857418927322757
iteration: 13, delta: 0.17939168383177906
iteration: 14, delta: 0.1565673276718128
iteration: 15, delta: 0.1348070775912846
iteration: 16, delta: 0.1419357513035936
iteration: 17, delta: 0.10725067960822193
iteration: 18, delta: 0.09543204437077818
iteration: 19, delta: 0.09109425895650142
iteration: 20, delta: 0.09380312801099225
iteration: 21, delta: 0.094411764481

C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C

### Step 4.6: (0% - we do it for you!)
 We did everything using the train and validation sets only until now. This enables us to choose the best hyperparameters `p_pca`, `K` and `C`.

 Looking at the performance/accuracy results you obtained in the previous step, answer the question: which `p_pca`, `K` and `C` combination gave the best validation set performance?
 AND: using the boaw encoding for that `p_pca, K` combination and the best C, finally apply the setting on the test set.

In [221]:
# NOT GRADED, BUT YOU CAN PLAY WITH THE VALUES OF K, p_pca AND C
# Best values:
K = 512
p_pca = 130
C = 0.001

# 1. Now we combine the training and validation, retrain a model with best C and
# best boaw representation (i.e. best K, p_pca) then predict on the test set.

# Generate boaw representation for this p_pca and K:
X_boaw_all = X_boaw_all_dict[(p_pca, K)]

# Split boaw data
X_boaw_train, X_boaw_val, X_boaw_test = split_ravdess_filewise_boaw(
    X_boaw_all, mask_train, mask_val, mask_test
)


# join train and val in one
X_boaw_trainval = np.concatenate([X_boaw_train, X_boaw_val])
Y_trainval = np.concatenate([Y_train, Y_val])

# 2. Now we run the classification model
clf = LinearSVC(C=C, random_state=1)
clf.fit(X_boaw_trainval, Y_trainval)
preds_test = clf.predict(X_boaw_test)
performance = np.mean(Y_test == preds_test)

print(f"\nFor K={K}, p_pca={p_pca} and C={C}:")
print(f"\tAccuracy: {performance}.")


For K=512, p_pca=130 and C=0.001:
	Accuracy: 0.42045454545454547.


C:\Users\Thijmen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(


Note: If everything went well, you should have an accuracy around 0.43 (on the test set) when using the best hyper-parameters `K`, `p_pca` and `C`.

**Before you submit, click Kernel > Restart & Run All to make sure you submit a working version of your code!**

--- End of the assignment ---